### PR TITLE
Add Client Subscriptions tests

### PR DESCRIPTION
**Pull request by karlnaden**
# Summary

Add Subscription notifications and other tests to the PAS client suite. This includes two primary updates to the Client Suite
1. Addition of a PAS Subscription Setup group where the client demonstrates its ability to subscribe to updates to PAS claims.
2. Update of the pending workflow tests to send a notification that the client system will react to.

To make that work and also make it flexible, several other changes were made:
1. Add tags for each workflow so tagging can be used rather than make/uses_requests
2. Add logic to create default pended and denied responses, as well as a default notification.
3. Allow testers to provide optional response bodies for all messages sent by Inferno

# Testing Guidance

- Documentation and the client postman collection have been updated to support a postman-based demo. In this case, the only outbound messages sent by the client tests are notifications which get routed to the [Subscriptions RI](https://subscriptions.argo.run/) which can receive notifications. This should be used to test
  - The tests passing with Inferno-generated responses.
  - The tests passing with tester-provided responses.
  - Other combinations hitting edge cases, like only some responses provided and malformed responses.
- Unit tests have not been updated, but currently still pass since there weren't any client unit tests to start with.Documentation and the client postman collection have been updated to support a postman-based demo. In this case, the only outbound messages sent by the client tests are notifications which get routed to the [Subscriptions RI](https://subscriptions.argo.run/) which can receive notifications. This should be used to testThe tests passing with Inferno-generated responses.The tests passing with tester-provided responses.Other combinations hitting edge cases, like only some responses provided and malformed responses.Unit tests have not been updated, but currently still pass since there weren't any client unit tests to start with.

### CODE REVIEW

File: lib/davinci_pas_test_kit/endpoints/subscription_read_endpoint.rb
Comment (tstrass) [inline lib/davinci_pas_test_kit/endpoints/subscription_read_endpoint.rb:1]: This looks to be identical to the one in the subscriptions test kit. Can we just reuse it instead of duplicating in this case?

File: lib/davinci_pas_test_kit/endpoints/subscription_status_endpoint.rb
Comment (tstrass) [inline lib/davinci_pas_test_kit/endpoints/subscription_status_endpoint.rb:1]: This looks to be identical to the one in the subscriptions test kit. Can we just reuse it instead of duplicating in this case?

File: lib/davinci_pas_test_kit/generator/group_generator.rb [hunk 1]
```diff
@@ -380,12 +380,13 @@ def description
...
               Demonstrate a complete prior authorization workflow including a period
-              during which the final decision is pending. This includes
+              during which the final decision is pending. This includes demonstrating
+              the ability of the server to
 
-              - Demonstrating the ability of the server to respond to a prior
-                authorization request with an `pended` status.
-              - Demonstrating the ability of the server to respond to a subsequent
-                inquiry request with a final decision for the request
+              - Respond to a prior authorization request with an `pended` status.
```
Comment (tstrass) [inline lib/davinci_pas_test_kit/generator/group_generator.rb:386]:
```suggestion
              - Respond to a prior authorization request with a `pended` status.
```

**Review by tstrass — CHANGES_REQUESTED** [review]

File: lib/davinci_pas_test_kit/custom_groups/v2.0.1/client_tests/pas_client_pas_inquire_request_bundle_validation_test.rb
[the remark this answers is not in the document]
Comment (karlnaden) [reply lib/davinci_pas_test_kit/custom_groups/v2.0.1/client_tests/pas_client_pas_inquire_request_bundle_validation_test.rb:1]: It is no longer generated because there are not enough of them to justify the complexity of creating a generator.  I can fix the file name / class name mismatch

File: lib/davinci_pas_test_kit/custom_groups/v2.0.1/client_tests/pas_client_subscription_pas_conformance_test.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            `http://hl7.org/fhir/us/davinci-pas/SubscriptionTopic/PASSubscriptionTopic`
+            in the `Subscription.criteria` element.
+          ))
+        end
+
+        filter_criteria = subscription.dig('_criteria', 'extension')
+          &.select do |ext|
+            ext['url'] == 'http://hl7.org/fhir/uv/subscriptions-backport/StructureDefinition/backport-filter-criteria'
+          end
+        unless filter_criteria&.length == 1
```
[the remark this answers is not in the document]
Comment (karlnaden) [reply lib/davinci_pas_test_kit/custom_groups/v2.0.1/client_tests/pas_client_subscription_pas_conformance_test.rb:37]: We still send notifications based on the Subscription because we can't really evaluate whether the criteria are met or not since they aren't formally defined in R4.  So I think that 201 created is still right. I'm open to a larger discussion about whether we should be more picky about what Subscriptions we accept, but I think this is fine for this initial PR with subscriptions.

File: config/PAS Client Test Suite Demo.postman_collection.json [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+						}
+					},
+					"response": []
+				}
+			]
+		}
+	],
+	"auth": {
+		"type": "oauth2",
+		"oauth2": [
```
[the remark this answers is not in the document]
Comment (karlnaden) [reply config/PAS Client Test Suite Demo.postman_collection.json:206]: When you get a new token, this setting automatically resets to on. I think this has to be fixed as a part of a larger improvement to our auth simulation, which is something I'm working on.

File: lib/davinci_pas_test_kit/docs/client_suite_description_v201.md [hunk 4]
```diff
@@ -47,67 +43,204 @@ validated with the Java validator using `tx.fhir.org` as the terminology server.
...
+   to see both passing and failing tests.
+1. Once Inferno finishes evaluating the requests, the test will complete allowing you to review the
+   results, including warning and error messages as well as requests associated with each test.
+
+#### Optional Demo Modifications
+
+This demo uses `id-only` notifications for Pended workflow. To see a demonstration of `full-resource`
+notifications, replace the string `id-only` in the "Create Subscription Request" entry under the 
+"Subscription Setup" folder in the collection with the string `full-resource` (found in an extension
+under the `_payload` element).
```
[the remark this answers is not in the document]
Comment (karlnaden) [reply lib/davinci_pas_test_kit/docs/client_suite_description_v201.md:145]: Correct - PAS doesn't support empty notifications. From the [spec](https://hl7.org/fhir/us/davinci-pas/STU2/specification.html#subscription):
> PAS Clients and Intermediaries SHALL support subscriptions with content=’id-only’ but MAY by mutual agreement and with appropriate security arrangements in place for push notifications containing PHI also support content=’full-resource’.

There's no mention of Empty, which I interpret to mean no support. From a practical perspective on empty, there's no way to know what Prior auth the notification is for.

File: lib/davinci_pas_test_kit/client_suite.rb [hunk 5]
```diff
@@ -46,6 +52,11 @@ class ClientSuite < Inferno::TestSuite
     suite_endpoint :post, TOKEN_PATH, TokenEndpoint
     suite_endpoint :post, SUBMIT_PATH, ClaimEndpoint
     suite_endpoint :post, INQUIRE_PATH, ClaimEndpoint
+    suite_endpoint :post, FHIR_SUBSCRIPTION_PATH, SubscriptionCreateEndpoint
+    suite_endpoint :get, FHIR_SUBSCRIPTION_INSTANCE_PATH, SubscriptionReadEndpoint
+    suite_endpoint :post, FHIR_SUBSCRIPTION_INSTANCE_STATUS_PATH, SubscriptionStatusEndpoint
+    suite_endpoint :get, FHIR_SUBSCRIPTION_INSTANCE_STATUS_PATH, SubscriptionStatusEndpoint
+    suite_endpoint :post, FHIR_SUBSCRIPTION_RESOURCE_STATUS_PATH, SubscriptionStatusEndpoint
```
[the remark this answers is not in the document]
Comment (karlnaden) [reply lib/davinci_pas_test_kit/client_suite.rb:61]: updated to fixed version

File: lib/davinci_pas_test_kit/client_suite.rb [hunk 6]
```diff
@@ -46,6 +52,11 @@ class ClientSuite < Inferno::TestSuite
     suite_endpoint :post, TOKEN_PATH, TokenEndpoint
     suite_endpoint :post, SUBMIT_PATH, ClaimEndpoint
     suite_endpoint :post, INQUIRE_PATH, ClaimEndpoint
+    suite_endpoint :post, FHIR_SUBSCRIPTION_PATH, SubscriptionCreateEndpoint
+    suite_endpoint :get, FHIR_SUBSCRIPTION_INSTANCE_PATH, SubscriptionReadEndpoint
+    suite_endpoint :post, FHIR_SUBSCRIPTION_INSTANCE_STATUS_PATH, SubscriptionStatusEndpoint
+    suite_endpoint :get, FHIR_SUBSCRIPTION_INSTANCE_STATUS_PATH, SubscriptionStatusEndpoint
```
[the remark this answers is not in the document]
Comment (karlnaden) [reply lib/davinci_pas_test_kit/client_suite.rb:60]: Inferno needs a notification bundle to build the status. This is a bug because I haven't updated it to work when Inferno generates the notification. Working on that now.

File: lib/davinci_pas_test_kit/custom_groups/v2.0.1/client_tests/pas_client_subscription_pas_conformance_test.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (tstrass) [reply lib/davinci_pas_test_kit/custom_groups/v2.0.1/client_tests/pas_client_subscription_pas_conformance_test.rb:37]: Makes sense, I'm good with that

File: lib/davinci_pas_test_kit/endpoints/claim_endpoint.rb [hunk 7]
```diff
@@ -1,25 +1,54 @@
...
 
 module DaVinciPASTestKit
   class ClaimEndpoint < Inferno::DSL::SuiteEndpoint
+    include SubscriptionsTestKit::SubscriptionsR5BackportR4Client::SubscriptionSimulationUtils
+    include ResponseGenerator
+    include URLs
+
+    # override the one from URLs
+    def suite_id
+      :davinci_pas_client_suite_v201 # TODO: don't hard-code the id here...
```
[the remark this answers is not in the document]
Comment (karlnaden) [reply lib/davinci_pas_test_kit/endpoints/claim_endpoint.rb:15]: Made an attempt to pull this information from the request object of the triggering request, which I think will be better - thanks for noticing this reminder.

File: lib/davinci_pas_test_kit/endpoints/subscription_status_endpoint.rb
Comment (tstrass) [reply lib/davinci_pas_test_kit/endpoints/subscription_status_endpoint.rb:1]: @karlnaden looks like you added some logic such that it's no longer identical to the endpoint in the subscriptions test kit, meaning it has to exist separately, right?

**Review by tstrass — CHANGES_REQUESTED** [review]

File: lib/davinci_pas_test_kit/endpoints/subscription_status_endpoint.rb
Comment (karlnaden) [reply lib/davinci_pas_test_kit/endpoints/subscription_status_endpoint.rb:1]: correct - I needed to be able to create the status even if no notification bundle was provided, so i think we keep this here. I could see a world in which we would factor the code in such a way that PAS could extend Subscriptions, but I don't think that's the right priority.

**Review by tstrass — APPROVED** [review]
Big improvement to the suite!
